utils/src/config.rs

```diff
@@ -0,0 +1,713 @@
+        }
+    }
+}
+
+impl Default for ServerTimeout {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            heartbeat_interval: default_heartbeat_interval(),
+            wait_synced_timeout: default_server_wait_synced_timeout(),
+            retry_timeout: default_retry_timeout(),
+            rpc_timeout: default_rpc_timeout(),
+            candidate_timeout: default_candidate_timeout(),
+            follower_timeout_range: default_follower_timeout_range(),
+        }
+    }
+}
+
+/// Curp client timeout settings
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Eq, Getters)]
+pub struct ClientTimeout {
+    /// Curp client timeout settings
+    #[getset(get = "pub")]
+    #[serde(with = "cluster_duration_format", default = "default_client_timeout")]
+    timeout: ClusterDuration,
+
+    /// Curp cliet wait sync timeout
+    #[getset(get = "pub")]
+    #[serde(
+        with = "cluster_duration_format",
+        default = "default_client_wait_synced_timeout"
+    )]
+    wait_synced_timeout: ClusterDuration,
+}
+
+impl ClientTimeout {
+    /// Create a new client timeout
+    #[must_use]
+    #[inline]
+    pub fn new(timeout: ClusterDuration, wait_synced_timeout: ClusterDuration) -> Self {
+        Self {
+            timeout,
+            wait_synced_timeout,
+        }
+    }
+}
+
+impl Default for ClientTimeout {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            timeout: default_client_timeout(),
+            wait_synced_timeout: default_client_wait_synced_timeout(),
+        }
+    }
+}
+
+/// Cluster configuration object
+#[allow(clippy::module_name_repetitions)]
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Getters)]
+pub struct LogConfig {
+    /// Log file path
+    #[getset(get = "pub")]
+    path: PathBuf,
+    /// Log rotation strategy
+    #[getset(get = "pub")]
+    rotation: RotationConfig,
+    /// Log verbosity level
+    #[getset(get = "pub")]
+    level: LevelConfig,
+}
+
+impl LogConfig {
+    /// Generate a new `LogConfig` object
+    #[must_use]
+    #[inline]
+    pub fn new(path: PathBuf, rotation: RotationConfig, level: LevelConfig) -> Self {
+        Self {
+            path,
+            rotation,
+            level,
+        }
+    }
+}
+
+/// Xline log rotation strategy
+#[non_exhaustive]
+#[allow(clippy::module_name_repetitions)]
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Eq)]
+pub enum RotationConfig {
+    /// Rotate log file in every hour
+    Hourly,
+    /// Rotate log file every day
+    Daily,
+    /// Never rotate log file
+    Never,
+}
+
+impl FromStr for RotationConfig {
+    type Err = anyhow::Error;
+    #[inline]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "hourly" => Ok(RotationConfig::Hourly),
+            "daily" => Ok(RotationConfig::Daily),
+            "never" => Ok(RotationConfig::Never),
+            _ => Err(anyhow!(format!("Failed to parse {s} to RotationConfig"))),
```

Comment:
I guess you don't need to manually implement these. Serde provides utils: see https://serde.rs/container-attrs.html#rename_all. Maybe `rename_all` isn't required either. I remember serde will recognize lower-case automatically.

Reply:
That's because we need to be compatible with the command line arguments, and the clap asked me to implement the `FromStr` trait for these arguments, like `Xline --log_rotate=hourly`

Reply:
I have already implemented a `parse_rotation` for `RotationConfig`, so we don't need to implement `FromStr` anymore